File: engine/core/Sound.cpp
#include "Sound.h"

#include "AudioAsset.h"
#include "Game.h"
#include "Scene.h"

Sound::Sound(Scene* scene, CsRef csRef) {
	m_scene = scene;
	m_scene->audio()->AddSound(this, csRef);
}

Sound::~Sound() {
	m_scene->audio()->RemoveSound(this);
}

void Sound::asset(const AudioAsset* value) {
	if (m_TrySetAsset(value)) {
		auto assetRef = CppRefs::GetRef(m_asset);
		m_scene->game()->callbacks().soundSetAsset(csRef(), assetRef);
	}
}

bool Sound::m_TrySetAsset(const AudioAsset* value) {
	if (m_asset == value)
		return false;

	m_asset = value;

	if (m_asset == nullptr)
		m_soundInstance = nullptr;
	else
		m_soundInstance = m_CreateSoundInstance();

	return true;
}

void Sound::Use3D(bool value) {
	if (m_use3d == value)
		return;

	m_use3d = value;

	if (m_soundInstance == nullptr)
		return;

	m_soundInstance = m_CreateSoundInstance();
}

std::unique_ptr<DirectX::SoundEffectInstance> Sound::m_CreateSoundInstance() {
	if (m_use3d)
		return m_asset->soundEffect->CreateInstance(DirectX::SoundEffectInstance_Use3D);
	
	return m_asset->soundEffect->CreateInstance(DirectX::SoundEffectInstance_Default);
}

DEF_FUNC(Sound, SetAsset, void)(CppRef soundRef, CppRef assetRef) {
	auto* sound = CppRefs::ThrowPointer<Sound>(soundRef);
	auto* asset = CppRefs::GetPointer<AudioAsset>(assetRef);

	sound->m_TrySetAsset(asset);
}

DEF_PROP_GETSET(Sound, bool, loop);

DEF_PROP_GET(Sound, bool, IsStopped);
DEF_PROP_GET(Sound, bool, IsPlaying);
DEF_PROP_GET(Sound, bool, IsPaused);

DEF_FUNC(Sound, Play, void)(CppRef iSoundRef) {
	auto iSound = CppRefs::ThrowPointer<Sound>(iSoundRef);

	iSound->Play();
}

DEF_FUNC(Sound, Stop, void)(CppRef iSoundRef) {
	CppRefs::ThrowPointer<Sound>(iSoundRef)->Stop();
}

DEF_FUNC(Sound, Pause, void)(CppRef iSoundRef) {
	CppRefs::ThrowPointer<Sound>(iSoundRef)->Pause();
}

DEF_FUNC(Sound, Resume, void)(CppRef iSoundRef) {
	CppRefs::ThrowPointer<Sound>(iSoundRef)->Resume();
}

File: engine/mono/Engine/Components/Audio/Sound.cs
using EngineDll;

namespace Engine {

    [GUID("6ddb549a-e7cd-46db-ab1e-1a6d975a1208")]
    public sealed class Sound {

        public CsRef csRef { get; private set; }
        public CppRef cppRef { get; private set; }
        public bool IsDestroyed => CppLinked.GetObjectByRef(csRef) == null;

        public bool IsStopped => Dll.Sound.IsPaused_get(cppRef);
        public bool IsPlaying => Dll.Sound.IsPlaying_get(cppRef);
        public bool IsPaused => Dll.Sound.IsPaused_get(cppRef);

        public bool Loop {
            get => Dll.Sound.loop_get(cppRef);
            set => Dll.Sound.loop_set(cppRef, value);
        }

        [Close] public AudioAsset asset {
            get => m_asset;
            set {
                m_asset = value;
                Dll.Sound.SetAsset(cppRef, m_asset != null ? m_asset.cppRef : CppRef.NullRef);
            }
        }
        
        private AudioAsset m_asset = null;

        public Sound() {
            csRef = CppLinked.CreateCsRef(this);
            cppRef = Dll.Scene.CreateSound(Game.sceneRef, csRef); 
        }

        public Sound(AudioAsset _asset): this() {
            asset = _asset;
        }

        private Sound(CppRef soundRef) {
            csRef = CppLinked.CreateCsRef(this);
            cppRef = soundRef;
        }

        public void Destroy() {
            Dll.Scene.RemoveSound(cppRef);
        }

        public void Play() => Dll.Sound.Play(cppRef);
        public void Stop() => Dll.Sound.Stop(cppRef);
        public void Pause() => Dll.Sound.Pause(cppRef);
        public void Resume() => Dll.Sound.Resume(cppRef);

        public static CsRef cpp_CreateSound(CppRef cppRef) {
            return new Sound(cppRef).csRef;
        }

        public static void cpp_SetAsset(CsRef soundRef, CppRef assetRef) {
            var sound = CppLinked.GetObjectByRef(soundRef) as Sound;
            if (sound == null)
                return;

            AudioAsset asset = null;
            
            if (assetRef != CppRef.NullRef) {
                var assetId = Dll.AudioAsset.assetId_get(assetRef);

                asset = new AudioAsset();
                asset.Init(assetId, assetRef);
            }
            sound.m_asset = asset;
        }
    }
}
